File: odyssey-server/src/routes/entries.rs
use axum::extract::{Path, Query, State};
use axum::routing::{delete, get, post};
use axum::{Json, Router};
use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};

use crate::AppState;
use crate::UserScope;
use crate::auth::AuthUser;
use crate::error::AppError;
use crate::models::EntryResponse;
use crate::validation;

/// Pre-validated entry fields produced by the validation loop.
struct ValidatedEntry {
    entry_date: NaiveDate,
    created_at: DateTime<Utc>,
    updated_at: DateTime<Utc>,
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/entries", post(create_entries))
        .route("/entries", get(list_entries))
        .route("/entries/{date}", delete(delete_entry))
}

// --- POST /entries ---

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncUploadRequest {
    pub entries: Vec<EntryInput>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntryInput {
    pub entry_date: String,
    pub journal_entry: Option<String>,
    pub gratitude: Option<String>,
    pub win: Option<String>,
    pub tension: Option<String>,
    pub single_word_feeling: Option<String>,
    pub latitude: Option<String>,
    pub longitude: Option<String>,
    pub city: Option<String>,
    pub state: Option<String>,
    pub country: Option<String>,
    pub feeling: i32,
    pub sleep_quality: i32,
    pub feeling_color_hex: String,
    pub drinks: i32,
    pub step_count: Option<i32>,
    pub walking_distance_meters: Option<f64>,
    pub sleep_hours: Option<f64>,
    pub screen_time_seconds: Option<f64>,
    pub pickups: Option<i32>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncResponse {
    pub synced_at: String,
    pub count: usize,
}

async fn create_entries(
    State(state): State<AppState>,
    user: AuthUser,
    Json(body): Json<SyncUploadRequest>,
) -> Result<Json<SyncResponse>, AppError> {
    let scope = UserScope::new(&user, &state);
    let entries = &body.entries;

    if entries.is_empty() || entries.len() > 100 {
        return Err(AppError::Validation(
            "entries array must contain 1-100 items".to_string(),
        ));
    }

    // Validate each entry and collect parsed date/time fields
    let mut validated: Vec<ValidatedEntry> = Vec::with_capacity(entries.len());
    for entry in entries {
        validation::validate_entry_date(&entry.entry_date)?;
        let entry_date = validation::validate_entry_date_range(&entry.entry_date)?;
        let created_at = validation::validate_timestamp(&entry.created_at, "createdAt")?;
        let updated_at = validation::validate_timestamp(&entry.updated_at, "updatedAt")?;
        validation::validate_feeling(entry.feeling)?;
        validation::validate_sleep_quality(entry.sleep_quality)?;
        validation::validate_drinks(entry.drinks)?;

        // String length constraints — text fields: 20 KB max
        validation::validate_encrypted_field_length(
            &entry.journal_entry, "journalEntry", 20 * 1024,
        )?;
        validation::validate_encrypted_field_length(
            &entry.gratitude, "gratitude", 20 * 1024,
        )?;
        validation::validate_encrypted_field_length(
            &entry.win, "win", 20 * 1024,
        )?;
        validation::validate_encrypted_field_length(
            &entry.tension, "tension", 20 * 1024,
        )?;
        // singleWordFeeling: 1 KB max
        validation::validate_encrypted_field_length(
            &entry.single_word_feeling, "singleWordFeeling", 1024,
        )?;
        // Location fields: 1 KB max
        validation::validate_encrypted_field_length(&entry.city, "city", 1024)?;
        validation::validate_encrypted_field_length(&entry.state, "state", 1024)?;
        validation::validate_encrypted_field_length(&entry.country, "country", 1024)?;

        // Hex color format
        validation::validate_hex_color(&entry.feeling_color_hex)?;

        validated.push(ValidatedEntry {
            entry_date,
            created_at,
            updated_at,
        });
    }

    // Upsert user
    sqlx::query(
        "INSERT INTO users (clerk_user_id, last_sync_at) VALUES ($1, NOW())
         ON CONFLICT (clerk_user_id) DO UPDATE SET last_sync_at = NOW()",
    )
    .bind(&scope.user_id)
    .execute(&scope.pool)
    .await?;

    // Upsert entries using pre-validated date/time fields
    for (entry, v) in entries.iter().zip(validated.iter()) {
        sqlx::query(
            "INSERT INTO entries (
                user_id, entry_date, journal_entry, gratitude, win, tension,
                single_word_feeling, latitude, longitude, city, state, country,
                feeling, sleep_quality, feeling_color_hex, drinks,
                step_count, walking_distance_meters, sleep_hours,
                screen_time_seconds, pickups, created_at, updated_at
            ) VALUES (
                $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12,
                $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23
            )
            ON CONFLICT (user_id, entry_date) DO UPDATE SET
                journal_entry = EXCLUDED.journal_entry,
                gratitude = EXCLUDED.gratitude,
                win = EXCLUDED.win,
                tension = EXCLUDED.tension,
                single_word_feeling = EXCLUDED.single_word_feeling,
                latitude = EXCLUDED.latitude,
                longitude = EXCLUDED.longitude,
                city = EXCLUDED.city,
                state = EXCLUDED.state,
                country = EXCLUDED.country,
                feeling = EXCLUDED.feeling,
                sleep_quality = EXCLUDED.sleep_quality,
                feeling_color_hex = EXCLUDED.feeling_color_hex,
                drinks = EXCLUDED.drinks,
                step_count = EXCLUDED.step_count,
                walking_distance_meters = EXCLUDED.walking_distance_meters,
                sleep_hours = EXCLUDED.sleep_hours,
                screen_time_seconds = EXCLUDED.screen_time_seconds,
                pickups = EXCLUDED.pickups,
                updated_at = EXCLUDED.updated_at",
        )
        .bind(&scope.user_id)
        .bind(v.entry_date)
        .bind(&entry.journal_entry)
        .bind(&entry.gratitude)
        .bind(&entry.win)
        .bind(&entry.tension)
        .bind(&entry.single_word_feeling)
        .bind(&entry.latitude)
        .bind(&entry.longitude)
        .bind(&entry.city)
        .bind(&entry.state)
        .bind(&entry.country)
        .bind(entry.feeling)
        .bind(entry.sleep_quality)
        .bind(&entry.feeling_color_hex)
        .bind(entry.drinks)
        .bind(entry.step_count)
        .bind(entry.walking_distance_meters)
        .bind(entry.sleep_hours)
        .bind(entry.screen_time_seconds)
        .bind(entry.pickups)
        .bind(v.created_at)
        .bind(v.updated_at)
        .execute(&scope.pool)
        .await?;
    }

    // Log to sync_log
    sqlx::query(
        "INSERT INTO sync_log (user_id, entries_pushed, entries_pulled) VALUES ($1, $2, 0)",
    )
    .bind(&scope.user_id)
    .bind(entries.len() as i32)
    .execute(&scope.pool)
    .await?;

    let synced_at = Utc::now()
        .to_rfc3339_opts(chrono::SecondsFormat::Millis, true);

    Ok(Json(SyncResponse {
        synced_at,
        count: entries.len(),
    }))
}

// --- GET /entries ---

#[derive(Debug, Deserialize)]
pub struct ListEntriesParams {
    pub since: Option<String>,
    pub cursor: Option<String>,
    pub limit: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ListEntriesResponse {
    pub entries: Vec<EntryResponse>,
    pub cursor: Option<String>,
}

async fn list_entries(
    State(state): State<AppState>,
    user: AuthUser,
    Query(params): Query<ListEntriesParams>,
) -> Result<Json<ListEntriesResponse>, AppError> {
    let scope = UserScope::new(&user, &state);
    let limit: i64 = params
        .limit
        .as_deref()
        .and_then(|l| l.parse::<i64>().ok())
        .unwrap_or(200)
        .min(200);

    // cursor takes priority over since
    let since_filter = params.cursor.as_deref().or(params.since.as_deref());

    let rows = if let Some(since) = since_filter {
        let since_dt = since
            .parse::<DateTime<Utc>>()
            .map_err(|e| AppError::Validation(format!("Invalid since/cursor timestamp: {e}")))?;

        sqlx::query_as::<_, crate::models::Entry>(
            "SELECT * FROM entries WHERE user_id = $1 AND updated_at > $2
             ORDER BY updated_at ASC LIMIT $3",
        )
        .bind(&scope.user_id)
        .bind(since_dt)
        .bind(limit)
        .fetch_all(&scope.pool)
        .await?
    } else {
        sqlx::query_as::<_, crate::models::Entry>(
            "SELECT * FROM entries WHERE user_id = $1
             ORDER BY updated_at ASC LIMIT $2",
        )
        .bind(&scope.user_id)
        .bind(limit)
        .fetch_all(&scope.pool)
        .await?
    };

    let next_cursor = if rows.len() as i64 == limit {
        rows.last().map(|e| {
            e.updated_at
                .to_rfc3339_opts(chrono::SecondsFormat::Millis, true)
        })
    } else {
        None
    };

    let entries: Vec<EntryResponse> = rows.into_iter().map(EntryResponse::from).collect();

    Ok(Json(ListEntriesResponse {
        entries,
        cursor: next_cursor,
    }))
}

// --- DELETE /entries/:date ---

#[derive(Debug, Serialize)]
pub struct DeleteResponse {
    pub deleted: bool,
}

async fn delete_entry(
    State(state): State<AppState>,
    user: AuthUser,
    Path(date): Path<String>,
) -> Result<Json<DeleteResponse>, AppError> {
    let scope = UserScope::new(&user, &state);
    sqlx::query("DELETE FROM entries WHERE user_id = $1 AND entry_date = $2")
        .bind(&scope.user_id)
        .bind(&date)
        .execute(&scope.pool)
        .await?;

    Ok(Json(DeleteResponse { deleted: true }))
}

File: odyssey-api/src/routes/entries.ts
import { Hono } from "hono";
import { eq, and, gt } from "drizzle-orm";
import { entries, users, syncLog } from "../db/schema";
import { syncUploadSchema, validateSyncEntry } from "../utils/validation";
import type { Env } from "../types";
import { createDb } from "../db/client";

const app = new Hono<{ Bindings: Env; Variables: { userId: string } }>();

// POST /entries — Batch upsert entries
app.post("/", async (c) => {
  const userId = c.get("userId");
  const body = await c.req.json();

  const parsed = syncUploadSchema.safeParse(body);
  if (!parsed.success) {
    return c.json({ error: "Validation failed", details: parsed.error.flatten() }, 400);
  }

  const db = createDb(c.env.DATABASE_URL);

  // Ensure user exists (upsert)
  await db
    .insert(users)
    .values({ clerkUserId: userId })
    .onConflictDoUpdate({
      target: users.clerkUserId,
      set: { lastSyncAt: new Date() },
    });

  const syncedAt = new Date().toISOString();
  let upsertedCount = 0;

  for (const entry of parsed.data.entries) {
    validateSyncEntry(entry);

    await db
      .insert(entries)
      .values({
        userId,
        entryDate: entry.entryDate,
        journalEntry: entry.journalEntry ?? null,
        gratitude: entry.gratitude ?? null,
        win: entry.win ?? null,
        tension: entry.tension ?? null,
        singleWordFeeling: entry.singleWordFeeling ?? null,
        latitude: entry.latitude ?? null,
        longitude: entry.longitude ?? null,
        city: entry.city ?? null,
        state: entry.state ?? null,
        country: entry.country ?? null,
        feeling: entry.feeling,
        sleepQuality: entry.sleepQuality,
        feelingColorHex: entry.feelingColorHex,
        drinks: entry.drinks,
        stepCount: entry.stepCount ?? null,
        walkingDistanceMeters: entry.walkingDistanceMeters ?? null,
        sleepHours: entry.sleepHours ?? null,
        screenTimeSeconds: entry.screenTimeSeconds ?? null,
        pickups: entry.pickups ?? null,
        createdAt: new Date(entry.createdAt),
        updatedAt: new Date(entry.updatedAt),
      })
      .onConflictDoUpdate({
        target: [entries.userId, entries.entryDate],
        set: {
          journalEntry: entry.journalEntry ?? null,
          gratitude: entry.gratitude ?? null,
          win: entry.win ?? null,
          tension: entry.tension ?? null,
          singleWordFeeling: entry.singleWordFeeling ?? null,
          latitude: entry.latitude ?? null,
          longitude: entry.longitude ?? null,
          city: entry.city ?? null,
          state: entry.state ?? null,
          country: entry.country ?? null,
          feeling: entry.feeling,
          sleepQuality: entry.sleepQuality,
          feelingColorHex: entry.feelingColorHex,
          drinks: entry.drinks,
          stepCount: entry.stepCount ?? null,
          walkingDistanceMeters: entry.walkingDistanceMeters ?? null,
          sleepHours: entry.sleepHours ?? null,
          screenTimeSeconds: entry.screenTimeSeconds ?? null,
          pickups: entry.pickups ?? null,
          updatedAt: new Date(entry.updatedAt),
        },
      });

    upsertedCount++;
  }

  // Log sync
  await db.insert(syncLog).values({
    userId,
    entriesPushed: upsertedCount,
  });

  return c.json({ syncedAt, count: upsertedCount });
});

// GET /entries — Fetch entries (optionally since timestamp)
app.get("/", async (c) => {
  const userId = c.get("userId");
  const since = c.req.query("since");
  const limit = Math.min(parseInt(c.req.query("limit") ?? "200"), 200);

  const db = createDb(c.env.DATABASE_URL);

  const cursor = c.req.query("cursor");

  const conditions = [eq(entries.userId, userId)];
  if (cursor) {
    conditions.push(gt(entries.updatedAt, new Date(cursor)));
  } else if (since) {
    conditions.push(gt(entries.updatedAt, new Date(since)));
  }

  const results = await db
    .select()
    .from(entries)
    .where(and(...conditions))
    .limit(limit)
    .orderBy(entries.updatedAt);

  const mappedEntries = results.map((row) => ({
    entryDate: row.entryDate,
    journalEntry: row.journalEntry,
    gratitude: row.gratitude,
    win: row.win,
    tension: row.tension,
    singleWordFeeling: row.singleWordFeeling,
    latitude: row.latitude,
    longitude: row.longitude,
    city: row.city,
    state: row.state,
    country: row.country,
    feeling: row.feeling,
    sleepQuality: row.sleepQuality,
    feelingColorHex: row.feelingColorHex,
    drinks: row.drinks,
    stepCount: row.stepCount,
    walkingDistanceMeters: row.walkingDistanceMeters,
    sleepHours: row.sleepHours,
    screenTimeSeconds: row.screenTimeSeconds,
    pickups: row.pickups,
    createdAt: row.createdAt.toISOString(),
    updatedAt: row.updatedAt.toISOString(),
  }));

  const nextCursor =
    results.length === limit
      ? results[results.length - 1].updatedAt.toISOString()
      : null;

  return c.json({ entries: mappedEntries, cursor: nextCursor });
});

// DELETE /entries/:date — Delete a single entry
app.delete("/:date", async (c) => {
  const userId = c.get("userId");
  const entryDate = c.req.param("date");

  const db = createDb(c.env.DATABASE_URL);

  await db
    .delete(entries)
    .where(and(eq(entries.userId, userId), eq(entries.entryDate, entryDate)));

  return c.json({ deleted: true });
});

export default app;
